add polling of status(on or off and brightness)

JustAro01 opened the pull request:
hey BazaJayGree66,

In this PR, I added polling of the status of the cololight (brightness and on/off). I found out by using wireshark and their [razer integration](https://www.razer.com/chroma-workshop#--apps--look--Y29sb2xpZ2h0L2NvbG9saWdodA_-__-_). I have only tested with cololight firmware version of: 5.18.8.9 (I think the newest version) and I don't know if it will work with older versions.(hopefully you can test that). I hope you don't think my code quality is too bad.

Greetings Pim

JustAro01 committed:
add light status and brightness polling

will poll status(on or off) and brightness of the colo light

JustAro01 committed:
added timeout if light is offline

JustAro01 committed:
remove try and except 

let home assistant handle the try and except, so you can see in error log if light is offline

BazaJayGee66 commented:
Amazing find!

Will try and test and merge in this week.

BazaJayGee66 commented on custom_components/cololight/light.py:
```diff
@@ -268,7 +267,18 @@ async def async_added_to_hass(self):
             self._hs_color = last_state.attributes.get("hs_color")
-
+    
+    def update(self):
```
Should use async update method
`async def async_update(self):`

BazaJayGee66 commented on custom_components/cololight/light.py:
```diff
@@ -268,7 +267,18 @@ async def async_added_to_hass(self):
+        elif data[40] == 206:#0xce
+           self._on = False
+        else:
+            return #if value is not 0xcf(on) or 0xce(off) stop the update then dont change anything
```
Out of curiosity, what state can be returned other than on and off?

JustAro01 replied:
The only states that the lamp returns are on or off. The official cololight app does not support receiving the effect(if you change the effect and reopen the app, it will default to "savasana" effect). But in my testing sometimes the light will return something else.(I think if it receives a on or off command). Thats why I made that else statement.

BazaJayGee66 replied:
I have been doing some testing, and have found that the response can return either state/brightness or effect/colour (this is why `data[40]` sometimes returns something other than state (on/off)).

However I haven't been able to get it to consistently return either state/brightness or effect/colour.
Sometimes by calling twice, you can get both (whereby the first call returns  state/brightness, and then the second call returns effect/colour).

BazaJayGee66 commented on custom_components/cololight/light.py:
```diff
@@ -196,8 +196,7 @@ def is_on(self):
     def should_poll(self) -> bool:
-        """Turn off, as cololight doesn't share state info"""
-        return False
+        return True
```
Looks like the default scan interval is 30s.
Think this is reasonable for this device, or should we increase it?

BazaJayGee66 replied:
Having a look what other components have done, feeling 30s poll is reasonable.

JustAro01 commented:
I did some more testing. There is currently a bug that if you turn the light on or off from home assistant. It will directly after that ask a update from the lamp. But because the light has not got the time to turn off or on it will give the old state. Do you know how to skip the update function after the turn on or off function has been called? Or is there a way to delay that update?

JustAro01 committed:
replace update to async_update & removed async_write_ha_state()

replace update to async_update and removed async_write_ha_state() because should_poll is enabled and according to home assistant docs: https://developers.home-assistant.io/docs/integration_fetching_data#push-vs-poll

BazaJayGee66 commented:
> Do you know how to skip the update function after the turn on or off function has been called? Or is there a way to delay that update?

I had noticed the same, but wasn't able to find if the update can be skipped.
One solution could be to add a delay in the update, with something like
```
await asyncio.sleep(0.5)
```
Not the most elegant solution, and not sure how it would handle if the light was repeatably turn on/off quickly.

Commit:
added skip after changing state on or off

And removed timeout of 4 seconds and let home assistant handle that. Also already changed manifest version if we forget to change it

JustAro01 commented:
I now added a bool var that is set to false if you turn the light on or off. So that the update function is skipped when turning the light on or off. What do you think about that solution? What also a solution is https://github.com/home-assistant/core/blob/191230f535b053ec76ae1164035de8879dfc2750/homeassistant/components/influxdb/sensor.py#L283 adding a throttle to the update function(I have not tested that or implented that). What solution do you prefer?

BazaJayGee66 commented:
I think having update as a boolean works.

Apologies, haven't had chance to test comprehensively yet.

BazaJayGee66 commented:
When running in Home Assistant, the response data from the socket get delayed over time.

Doing some investigation, it looks like the data returned is not the latest state, but rather the last issued command in sequence.

I believe this might be due to how we're using the socket library.


Here's a simple example, where the output isn't the latest state (80 brightness), but what the brightness was in sequence for each command: 
```python
from cololight.light import (
    PyCololight,
    ColourException,
    ColourSchemeException,
    ModeExecption,
    CycleSpeedException,
)
import time

light = PyCololight("x.x.x.x")


def status():
    light._sock.sendto(
        bytes.fromhex(
            "535a303000000000001e000000000000000000000000000000000200000000000000000003020101"
        ),
        (light.host, light.port),
    )
    data = light._sock.recvfrom(4096)[0]
    print(data[41])
    time.sleep(0.5)


light.on = 10
time.sleep(0.5)
light.on = 20
time.sleep(0.5)
light.on = 30
time.sleep(0.5)
light.on = 40
time.sleep(0.5)
light.on = 30
time.sleep(0.5)
light.on = 60
time.sleep(0.5)
light.on = 70
time.sleep(0.5)
light.on = 80
time.sleep(0.5)

status()
status()
status()
status()
status()
status()
status()
status()

### Output
# 10
# 20
# 30
# 40
# 30
# 60
# 70
# 80
```

BazaJayGee66 committed:
Abstract update state to be a method under PyCololight class

BazaJayGee66 committed:
Update to close socket connection after sending or receiving data

BazaJayGee66 commented:
Have updated to close the socket  connection after sending/receiving data.

BazaJayGee66 committed:
Add changelog

Commit:
fixed crashing ha core if device is offiline

JustAro01 commented:
The code works great. Only issue what I had was if the device was offline. Home assistant core crashes. Because it will infinity wait for the message of the light. (I had removed that timeout because I thought that home assistant handles that). I have set the timeout to 4 seconds.

BazaJayGee66 committed:
Make device unavailable if communication times out

BazaJayGee66 commented:
Good find, have updated to make device unavailable if it times out.
